vivisect/analysis/elf/elfplt_late.py

```diff
@@ -0,0 +1,376 @@
+    for fva in vw.getFunctions():
+        if pltva <= fva < (pltva+pltsz) and fva not in curpltset:
+            logger.debug("existing PLT Function: 0x%x", fva)
+            curpltset.add(fva)
+
+    curplts = list(curpltset)
+    curplts.sort()
+
+    if len(curplts):    # make sure we have *any* functions
+        # just make the first thing a function.  this should take care of any LazyLoader
+        # or at the very least, the first PLT entry.  it should also keep us from backing
+        # up too far during PLT-Func-Distance algorithm.
+        logger.debug('analyzePLT(0x%x, 0x%x): making first location the PLT into a function', pltva, pltsz)
+        vw.makeFunction(pltva)
+
+    # now figure out the distance from function start to the GOT xref 
+    # and between PLT functions
+    lastva = pltva
+    jmpheur = {}
+    distanceheur = {}
+    for fva in curplts:
+        fsz = vw.getFunctionMeta(fva, 'Size')
+        gotva, gotsz = elfplt.getGOT(vw, fname)
+
+        offset = 0
+        while offset < fsz:
+            locva, lsz, ltype, ltinfo = vw.getLocation(fva + offset)
+            xrefsfrom = vw.getXrefsFrom(locva)
+            for xrfr, xrto, xrtype, xrtinfo in xrefsfrom:
+                if gotva <= xrto < gotva+gotsz:
+                    offcnt = jmpheur.get(offset, 0)
+                    jmpheur[offset] = offcnt + 1
+
+            offset += lsz
+
+        # capture distance between functions
+        delta = fva - lastva
+        distanceheur[delta] = distanceheur.get(delta, 0) + 1
+        lastva = fva
+
+    if None not in (gotva, gotsz):
+        logger.debug("GOT/size: 0x%x/0x%x", gotva, gotsz)
+
+    # GOT-XREF-Offset method
+    if len(jmpheur):
+        algo_GOT_XREF_Offset(vw, jmpheur, pltva, pltsz)
+
+    else:
+        logging.info("analyzePLT(0x%x, 0x%x) skipping GOT-XREF-Offset method: no existing functions found", pltva, pltsz)
+
+    # PLT-Func-Distance method
+    if len(distanceheur):
+        algo_PLT_Func_Distance(vw, curplts, distanceheur, pltva, pltsz)
+
+    else:
+        logging.info("skipping analyzePLT(0x%x, 0x%x) (PLT-Func-Distance method): no existing functions found", pltva, pltsz)
+
+    logger.info("elfplt_late (done): pltva: 0x%x, %d", pltva, pltsz)
+
+
+def algo_GOT_XREF_Offset(vw, jmpheur, pltva, pltsz):
+    # this seems to work for many PLT's, but only if the xref is identifiable 
+    # without function analysis.  it fails on i386-pic code which uses ebx 
+    # (which is handed into the call as a base address)
+
+    # if we have what we need... scroll through all the non-functioned area
+    # looking for GOT-xrefs
+    offbycnt = [(cnt, off) for off, cnt in jmpheur.items()]
+    offbycnt.sort(reverse=True)
+    logger.debug("PLT Segment Data: %r", vw.getSegment(pltva))
+    cnt, realoff = offbycnt[0]
+
+    # now roll through the PLT space and look for GOT-references from 
+    # locations that aren't in a function
+    logger.info("... analysis mode 1: GOT-XREF Offset")
```

Comment:
Is there a more descriptive name than `analysis mode 1: GOT-XREF Offset`? It sounds like it's calculating a distance metric almost to an initial read, but I'm not entirely sure of that?

Reply:
not really for a name, and probably not anything worth putting in a log message.

Reply:
updated.